feat: add `_comp_xfunc`

akinomyoga opened the pull request:
Resolve #732

- 2c7ca225 add `_comp_xfunc`
- 2b4bb190 use `_comp_xfunc` instead of `_xfunc`
- fc0b4b87 add `_comp_deprecate_func` for backward-compatibility names

The function `_comp_deprecate_func` is introduced to later automatically detect the use of old function names for #542. To implement #542, some codes can be added in the automatically generated function body in `_comp_deprecate_func`.

akinomyoga commented on bash_completion:
```diff
@@ -2432,6 +2457,28 @@ _xfunc()
+_comp_deprecate_func _ssh_identityfile _comp_xfunc_ssh_identityfile
+_comp_deprecate_func _ssh_options _comp_xfunc_ssh_options
+_comp_deprecate_func _ssh_query _comp_xfunc_ssh_query
+_comp_deprecate_func _ssh_suboption_check _comp_xfunc_ssh_suboption_check
```
@scop has suggested in https://github.com/scop/bash-completion/issues/732#issuecomment-1094831476 that the backward-compatibility names might be defined in a separated file (e.g. in `/etc/bash_completion.d`), but currently these compatibility names are tentatively included in `bash_completion`.
- If we put the file in `/etc/bash_completion.d`, something unwanted might occur when the user sets the custom `BASH_COMPLETION_COMPAT_DIR`, so it might be better to be in another directory such as `helpers`?
- Another option is to put these compatibility names in the same file as the new API. For example, define the compatibility name in _apt_cache_packages.

akinomyoga replied:
I now noticed that this causes the problem when the new name is not yet defined when the compatibility name is used. So, the compatibility name should be defined after the new name is defined (for example in the same file as the new name), or the function of the compatibility name should call the function of the new name through `_comp_xfunc`.

scop commented on test/t/unit/test_unit_xfunc.py:
```diff
@@ -0,0 +1,50 @@
+            bash_env.write_variable(
+                "BASH_COMPLETION_USER_DIR", "$PWD/_comp_xfunc", quote=False
+            )
+
```
Similarly here, perhaps assert that `_comp_xfunc_non_standard_name` is not yet defined?

akinomyoga replied:
Yes! I will process it in the other [(conversation)](https://github.com/scop/bash-completion/pull/734/files#r848820292),

akinomyoga replied:
573e0db0

scop commented on bash_completion:
```diff
@@ -2419,6 +2419,31 @@ _completion_loader()
+}
+
+#---------------------------------------
+# The following functions are deprecated / backward-compatible names
```
I'd suggest adding a `@deprecated` to each deprecated function's commentary instead (see existing examples) -- that way I think people are more likely to notice than if/when this section becomes longish.

akinomyoga replied:
Thanks! I haven't noticed the existing ones.

akinomyoga replied:
2585250f

scop commented on bash_completion:
```diff
@@ -2432,6 +2457,8 @@ _xfunc()
     "$@"
 }
 
+#---------------------------------------
```
This can likely go if we add per function `@deprecated` markers.

akinomyoga replied:
2585250f

scop commented on completions/cvs:
```diff
@@ -1,5 +1,10 @@
 # cvs(1) completion                                        -*- shell-script -*-
 
+_comp_xfunc_cvs_roots()
+{
+    _cvs_roots
+}
+
```
Should we not rename + `_comp_deprecate_func` `_cvs_roots` to this instead?

akinomyoga replied:
I have kept the original function `_cvs_roots` for internal use by `cvs` and just added an interface here because there seem to be many other functions similar to `_cvs_roots` below (i.e., _cvs_entries, _cvs_modules, _cvs_commands. etc.)

- Nevertheless, I may directly rename only `_cvs_roots` in the subsequent code.
- Another option is to make all the similar functions *xfunc* for consistency.

There are similar cases also in several other files.

scop replied:
Hmm, I see. But then again all cases except this and ssh in this PR take the rename+deprecate approach. None of the choices strikes me obviously better than all others, but I think I'd prefer the consistency by not leaving wrappers like this behind, and making things xfunc only when we know we have xfunc use for them from somewhere else.

akinomyoga replied:
OK! So, if I don't misunderstand, ccf13887?

scop replied:
Yes, that's what I meant, thanks.

scop commented on completions/ssh:
```diff
@@ -1,8 +1,38 @@
 # ssh(1) completion                                        -*- shell-script -*-
 
+_comp_xfunc_ssh_query()
+{
+    _ssh_query "$@"
+}
+
+_comp_xfunc_ssh_options()
+{
+    _ssh_options "$@"
+}
+
+_comp_xfunc_ssh_suboption_check()
+{
+    _ssh_suboption_check "$@"
+}
+
+_comp_xfunc_ssh_identityfile()
+{
+    _ssh_identityfile "$@"
+}
+
+_comp_xfunc_ssh_scp_remote_files()
+{
+    _scp_remote_files "$@"
+}
+
+_comp_xfunc_ssh_scp_local_files()
+{
+    _scp_local_files "$@"
+}
+
```
Should we not rename + `_comp_deprecated_func` the originals instead?

akinomyoga replied:
I will treat them the same as `_cvs_roots` after the treatment of `_cvs_roots` is settled.

scop commented on bash_completion:
```diff
@@ -2419,6 +2419,31 @@ _completion_loader()
+# `_comp_xfunc_$1_$2' is used for the actual name of the shell function.
+_comp_xfunc()
+{
+    local __xfunc_name=$2
```
Cosmetic, but I suppose we could use `xfunc_name` for this local variable's name here, no need for the underscores?

akinomyoga replied:
Yeah, this is actually related to https://github.com/scop/bash-completion/discussions/539#discussioncomment-2552484. But in this case, there is much less possibility of the variable name clash. I'll just follow your suggestion! Thanks.

akinomyoga replied:
c056c2d8

scop reviewed (approved):
Looks good to me, thanks!

Left a bunch of rather trivial comments, pre-approved with those addressed, but feel free to bounce back if you think they deserve more discussion or I missed something.

akinomyoga commented on test/t/unit/test_unit_xfunc.py:
```diff
@@ -0,0 +1,50 @@
+            bash_env.write_variable(
+                "BASH_COMPLETION_USER_DIR", "$PWD/_comp_xfunc", quote=False
+            )
+
```
Good suggestion!

akinomyoga replied:
573e0db0 I have added also the test that the function is defined before the second invocation. If that addition is not needed, please let me know.

akinomyoga commented on test/t/unit/test_unit_xfunc.py:
```diff
@@ -0,0 +1,50 @@
+            )
+            assert output.strip() == "util1[<a b><cde><fgh>]"
+
+            # first invocation (completion/xfunc-test1 is not sourced)
```
f623b9d8

akinomyoga commented on test/t/unit/test_unit_xfunc.py:
```diff
@@ -0,0 +1,50 @@
+            )
+            assert output.strip() == "util2[<a b><cde><fgh>]"
+
+            # first invocation (completion/xfunc-test2 is not sourced)
```
f623b9d8

akinomyoga commented:
I have rebased and squashed the changes. For the additional changes compared to the previous version, I have commented on the commit hashes in each conversation.

akinomyoga commented:
Now I noticed that there can be a case where an external completion calls any function that is not marked *xfunc* by us through `_xfunc`. Maybe we need to still allow such cases for backward compatibility as 35fbf57c, or do we recommend continuing to use the deprecated `_xfunc` in such a case? Hmm, what do you think?

scop commented:
35fbf57cbc59e9879e59bd19dc044b1bc5f43af0 makes sense to me. With that, couldn't we just `_comp_deprecate_func _xfunc _comp_xfunc` considering that our separate completion files don't have any functions that would _not_ start with an underscore?

akinomyoga commented:
Thank you!
- ce9b010a I have supplemented 35fbf57c with an appropriate update of the description
- ffb59f43 I have moved `_comp_deprecate_func` to the beginning of the file to use it for functions defined in the main `bash_completion`, and applied it to `_comp_xfunc`
- f8a6a096 I have added input checks to `_comp_deprecate_func`

----

I have rebased and squashed.

akinomyoga commented:
Thanks!
